### ISSUE TITLE
reproducible results

### ISSUE THREAD

**Issue by NamLQ** [rstudio/keras#42]
Hi everyone!

Each time I train a model on the same data set, I receive the different result. I already set the seed set.seed(1) but it does not work. How can I make the results the same for each run? 
Thank you very much!

```
library(keras)
# Generate dummy data
set.seed(1)
data <- matrix(runif(1000*100), nrow = 1000, ncol = 100)
labels <- matrix(round(runif(1000, min = 0, max = 1)), nrow = 1000, ncol = 1)

# repeat running the code from this:
model <- keras_model_sequential()

# add layers and compile the model
model %>% 
    layer_dense(units = 32, activation = 'relu', input_shape = c(100)) %>% 
    # layer_dropout(rate = 0.25) %>%
    # layer_batch_normalization() %>%
    layer_dense(units = 1, activation = 'relu') %>% 
    compile(
        optimizer = 'adam',
        loss = 'mean_squared_error',
        metrics = 'mean_squared_error'
    )

# Train the model, iterating on the data in batches of 32 samples
set.seed(1)
model %>% fit(data, labels, epochs=10, batch_size=32)
```

**Comment by terrytangyuan** [comment]
It's extremely hard to get a deterministic model since you'll have to control the random seeds properly in each Keras layer, e.g. layers like dropout, initializers, etc. have randomness built in. These will have to be controlled well in Keras' Python front-end. If you really want a very robust model, you really should try to take advantage of this randomness nature when train the model. Hope this helps. Closing this for now.

**Comment by topepo** [comment]
Can we reopen this? 

R (and R users) highly prize reproducibility. It is difficult to generate teaching materials, reports, books, and other items if we get different results on every run.  

I understand the issue about seeds in different places. Perhaps we could set them to something similar to `sample.int(10^5, 1)` so that in each location that a seed is required, it could pull from the global random number stream. In that way, we could set the seed for the session and the constituent layers, initializes, and other parts get a deterministic value for the seed.

**Comment by terrytangyuan** [comment]
Is this issue happening in Python API as well? If so, this should really be an issue there.

**Comment by jjallaire** [comment]
Here is what we have currently in the FAQ: https://keras.rstudio.com/articles/faq.html#how-can-i-obtain-reproducible-results-using-keras-during-development

If there is something we can do in the package that facilitates better reproducibility I'm all for it, just let me know what you think we need to do.

**Comment by terrytangyuan** [comment]
I think setting both R and TF seeds is the best we can probably do here.

**Comment by jjallaire** [comment]
There are some additional Python/NumPy random seeds set in this Keras
Python example:

https://keras.io/getting-started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development

Do we need to set those as well?

If so perhaps we should add a `session_with_seed` function that sets all of
these values and returns the session (invisibly, since users typically
won't need to compute on it). The `session_with` prefix is important b/c as
a side effect this also sets the TF session.

Actually, perhaps it should be `tf_session_with_seed` and the function
belongs in the tensorflow package?




On Sat, Sep 2, 2017 at 10:06 AM, Yuan (Terry) Tang <notifications@github.com

**Comment by terrytangyuan** [comment]
@jjallaire Sounds good to me! I took a stab at https://github.com/rstudio/tensorflow/pull/174 for this.   This will be useful for small computations that can run in single thread but unfortunately not too useful for multi-thread though. I do believe people nowadays really try take advantage of this non-reproducibility to create more robust models as well as achieve model parallelism. 

I would say give it a shot once that PR got merged. I think `np` and `random` should give a pretty good coverage of the functions in Keras that use random generators.

**Comment by topepo** [comment]
To get it from the `keras` side: rstudio/keras#119
